### PR TITLE
PrioritiesSprint3.md

### PR DESCRIPTION
**Pull request by tursynay**
Need @cbizon to review and edit

### CODE REVIEW

**Comment by tursynay** [comment]
Dear TACT team members, please review the Sprint 3 priorities and their associated tasks by the end of Thursday, April 25. If you have any concerns, post in comments.

Sprint 3 starts on Monday, April 29.

File: PrioritiesSprint3.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+# Develop Pathfinder Query
+## Pre-development work:
+- Design new landing page to distinguish types of interactions (templated query, pathfinder, multiquery(?)) (UI)
+- Design query selection interface - Dependencies: decisions on categories of nodes to be allowed as starting points and any other factors allowed for query selection (UI, ARAs)
+- User testing path grouping strategies (UI, ARAs)
+- Design results page path  and graph display (similar to results now) - Dependencies: understanding of the number of paths and how they will be grouped (UI, ARAs)
+- Research to identify new facets needed. e.g. facets for other categories - Dependencies: pathfinder data examples (UI, ARAs)
+- User research filter designs for new filters like metaedge type or other. (UI)
+- Provides example datasets to UI (ARAs)
```
Comment (andrewsu) [inline PrioritiesSprint3.md]: What sort of dataset is being requested here?

Comment (cbizon) [reply PrioritiesSprint3.md]: @uhbrar ?

File: PrioritiesSprint3.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+# Scalability
+## Eel (CI):
+- Sending Concurrent queries (MVP1&MVP2) in batch of 5,10,15,20,25,30 (Test Env)
+
+# CQS Workflows
+## Eel (CI):
+-  Test and iterate MVP1 workflows
+-  Push these workflows to TEST
+
+# Fix Failing Automated Tests
```
Comment (karafecho) [inline PrioritiesSprint3.md:54]: This priority is not clear to me.

File: PrioritiesSprint3.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+The current number of tests is 484, so the number of new tests is (Number of tests - 484)
```
Comment (maximusunc) [inline PrioritiesSprint3.md]: I'm not entirely sure who added the numbers here. But I do think we need to iron out both what the realistic goals should be, who is actually in charge of meeting them, and is there any consequence for not meeting the goals? (i.e. if an ARA has too many failing tests, what happens? Are they excluded from the ARS until they are in compliance?) Right now, we have some goals that no ARA is currently meeting, and from the Testing team perspective, I'm not really sure what I can do about that other than provide the report.

File: PrioritiesSprint3.md (hunk 2 above)
Comment (cbizon) [reply PrioritiesSprint3.md:54]: Is there something particular that doesn't make sense?

File: PrioritiesSprint3.md (hunk 3 above)
Comment (cbizon) [reply PrioritiesSprint3.md]: I added the numbers - We had a problem in earlier sprints where we set a target, but at the same time we added tests over the sprint, so the target got harder to achieve over the course of the sprint.  The numbers here were an attempt to constrain what is required to something that wouldn't change over the month.

The overall idea would be that ARAs and CQS would meet the targets which we would then progressively make more difficult over time.  So I think the Testing team is meeting what it needs to do, the impetus is on the ARAs at this point.

We have not really discussed what happens if not met and I agree that's a good topic for conversation.  My personal feeling at the moment is that being too draconian will be counter productive, but maybe we would need to be to get it to happen?

File: PrioritiesSprint3.md (hunk 2 above)
Comment (karafecho) [reply PrioritiesSprint3.md:54]: Yes, Max and I discussed it a bit (please see his comment). Essentially, I'm not sure where the numbers are coming from, and I think "Number of New Tests" should be specified.

File: PrioritiesSprint3.md (hunk 3 above)
Comment (maximusunc) [reply PrioritiesSprint3.md]: I think that sounds good. My suggestion then is to maybe switch up the objectives a little bit to give clearer concrete numbers? Here's a first proposal:
- ARAs must pass at least 350 tests
- ARAs must provide results for at least 425 tests
- The CQS must pass at least 100 tests

Granted, with the addition of new tests would probably help make these objectives easier. (Maybe that's also incentive for the consortium to add more test assets?)

Comment (cbizon) [reply PrioritiesSprint3.md]: Yeah, I'm really trying to unmix the ARA target from the changes that the testing group made, which makes it hard to provide concrete numbers

**Review by gglusman — APPROVED** [review]

File: PrioritiesSprint3.md (hunk 2 above)
Comment (cbizon) [reply PrioritiesSprint3.md:54]: Do you mean specified like given a number or specified like defined what the phrase "Number of New Tests" means?

Comment (karafecho) [reply PrioritiesSprint3.md:54]: I mean replace "Number of New Tests" with a specific number or delete it, just so folks know exactly what the target is. At present, there are over 600 test assets, so we could say something like:

For 600 test assets:
- Reduce Test Failures for ARAs to < 100
- Reduce No-Results for ARAs to < 50 
- Reduce Test Failures for CQS to < 20

I'm not sure what the "right" numbers should be, but it looks like you and Max are working that out.

Comment (cbizon) [reply PrioritiesSprint3.md:54]: Well this is exactly the problem: We know how many tests there today.  We don't know how many there will be in a month.

File: PrioritiesSprint3.md (hunk 1 above)
Comment (uhbrar) [reply PrioritiesSprint3.md]: The "dataset" is just the outputs for the 4 example queries.

File: PrioritiesSprint3.md (hunk 3 above)
Comment (CaseyTa) [reply PrioritiesSprint3.md]: Sorry if I'm using the wrong testing terminology, but is there a way to create a test suite of the current test assets and define our performance goals relative to just that test suite? Hopefully that way, when new tests are added, it doesn't change the goal posts.

File: PrioritiesSprint3.md (hunk 2 above)
Comment (karafecho) [reply PrioritiesSprint3.md:54]: Yes, but I was suggesting that we don't worry about the future and just pick a cap. Casey's suggestion below is similar in sentiment to mine but more clearly defined.

File: PrioritiesSprint3.md (hunk 3 above)
Comment (cbizon) [reply PrioritiesSprint3.md]: @maximusunc what do you think?

Comment (maximusunc) [reply PrioritiesSprint3.md]: That's an excellent idea and very easy to implement. Let's go that route and have some concrete numbers for this sprint goal.

**Review by bill-baumgartner — APPROVED** [review]

File: PrioritiesSprint3.md (hunk 3 above)
Comment (maximusunc) [reply PrioritiesSprint3.md]: So it looks like there are as of right now 604 tests. We can pin that and use it as the Sprint 3 Test Suite. So maybe we make the goals (I think I still like the numbers I posted above?):
- ARAs must pass at least 350 tests
- ARAs must provide results for at least 425 tests
- The CQS must pass at least 100 tests

Comment (cbizon) [reply PrioritiesSprint3.md]: Great, sounds good. I'll update the PR

Comment (cbizon) [reply PrioritiesSprint3.md]: OK, PR updated

File: PrioritiesSprint3.md (hunk 1 above)
Comment (cbizon) [reply PrioritiesSprint3.md]: So do we want everything  ARA to provide outputs for all 4 example queries?  Or an ARA or what?

**Review by maximusunc — APPROVED** [review]

**Review by karafecho — APPROVED** [review]

File: PrioritiesSprint3.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+## Octopus (TEST):
+- UI implements "treats" refactor including provenance (UI)
+
+# Scalability
+## Eel (CI):
+- Sending Concurrent queries (MVP1&MVP2) in batch of 5,10,15,20,25,30 (Test Env)
+
+# CQS Workflows
+## Eel (CI):
+-  Test and iterate MVP1 workflows
```
Comment (sierra-moxon) [inline PrioritiesSprint3.md:51]: Here are the details of "how to test MVP1 workflows with the CQS":
- run every CQS template through DEV WFR and confirm that no errors are produced.  Could be that there are no records returned, and that is ok. 
- make sure that each CQS template passes TRAPI JSON validation - maybe we can incorporate reasoner-validator tests into the CQS template repo, but if not, at least manually.
------- MERGE TEMPLATES INTO CI -------------
- Solicit at least 3 Q/A pairs from each group that has a CQS template
- Create a series of TestCases (and a TestSuite)
- Modify the TestRunner to run the CQS TestCases (and TestSuite) against the CI version of CQS (or at least against CI ARS?)?
- Assemble a list of PKs to examine manually from CI
- Examine each PK to see that SPO + support edges are filled out with the provenance we expect in CI
------- MERGE TEMPLATES INTO TEST----------

**Review by sierra-moxon — APPROVED** [review]

File: PrioritiesSprint3.md (hunk 4 above)
Comment (karafecho) [reply PrioritiesSprint3.md:51]: Here's the protocol for submitting new CQS templates: https://github.com/TranslatorSRI/CQS/tree/main/templates.

File: PrioritiesSprint3.md (hunk 3 above)
Comment (suihuang-ISB) [reply PrioritiesSprint3.md]: > 
> The overall idea would be that ARAs and CQS would meet the targets which we would then progressively make more difficult over time. So I think the Testing team is meeting what it needs to do, the impetus is on the ARAs at this point.
> 


Chris, this is important but I am not sure what we mean by "more difficult" - if it is based on the # of tests passed, then it is self-referential. My concern is: How will the current performance generalize beyond these MVP1/2 cases?  The test assets are very biased and do not cover a large variety of queries - they are quite dependent (derived) from each other .... Maybe for discussion of next testing in next sprint.

**Review by suihuang-ISB — APPROVED** [review]

**Review by GregHydeDartmouth — APPROVED** [review]

**Review by vdancik — APPROVED** [review]

**Review by CaseyTa — APPROVED** [review]

File: PrioritiesSprint3.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+-  Push these workflows to TEST
+
+# Fix Failing Automated Tests
+## Pre-Development
+- Implement MVP2 benchmarks (Testing)
+## Eel (CI):
+- Deploy validation test runner (Testing)
+- Deploy stress test runner (Testing)
+- ARAs must pass at least 350 tests in the Sprint 3 Test Suite (ARAs)
```
Comment (saramsey) [inline PrioritiesSprint3.md:60]: I have concerns about setting 350 as the requirement for within 30 days, not knowing the baseline of what proportion of tests are currently passing for ARAX, and not even knowing what the 600 tests are (I only see 402 tests listed in the Information Radiator at present). I'm also not sure what exactly counts as a "failed test". If any one ARA or if the CQS fails the test, is that counted as a failure for ARAX? What about time-outs?

**Review by saramsey — APPROVED** [review]

**Review by newgene — APPROVED** [review]

File: PrioritiesSprint3.md (hunk 5 above)
Comment (cbizon) [reply PrioritiesSprint3.md:60]: Hi @saramsey 
* The 350 is for each ARA independently
* @maximusunc  posts the results for each week to the slack channel translator-testing-results, where you can see the breakdown per ARA (not yet for this week, but I'm sure it's coming)
* If you look in the first column of what Max posted last week ago, ARAX passed 253.  But as you say, that is on a smaller number of tests
* In the most recent run ([View in the Information Radiator](https://informationradiator.renci.org/test-runs/55)) there are 600.   There should also be a place in github to find the definition of the test 3 suite, but I'm not 100% sure where it is.

**Pull request merged**